The initial commit for the Anaplan plugin (0.0.1-snapshot)

niceyang opened the pull request:
1. Anaplan library code from repo(https://github.com/anaplaninc/anaplan-java-client)
2. Wrapper class of the Anaplan service required by the plugins (AnaplanService.java)
3. Anaplan Function runner plugin
4. Anaplan Sink plugin
5. Anaplan Source plugin

* Anaplan Source plugin is built upon GCS plugin in repo(https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/gcs/source/GCSSource.java)
* GCSPath.java is copied for supporting the Anaplan Source plugin from repo (https://github.com/data-integrations/google-cloud/tree/develop/src/main/java/io/cdap/plugin/gcp/gcs)

niceyang committed:
The initial commit for the Anaplan plugin (0.0.1-snapshot), including:
1. Anaplan library code from repo(https://github.com/anaplaninc/anaplan-java-client)
2. Wrapper class of the Anaplan service required by the plugins (AnaplanService.java)
3. Anaplan Function runner plugin
4. Anaplan Sink plugin
5. Anaplan Source plugin

* Anaplan Source plugin is built upon GCS plugin in repo(https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/gcs/source/GCSSource.java)
* GCSPath.java is copied for supporting the Anaplan Source plugin from repo (https://github.com/data-integrations/google-cloud/tree/develop/src/main/java/io/cdap/plugin/gcp/gcs)

blueandgold commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+    String getPassword();
+  }
+
+  public static final String NAME_SERVICE_LOCATION = "serviceLocation";
```
What is all these `NAME_` prefixes?   The actual values don't have `name`.

niceyang replied:
NAME_ this is the naming convention learned from CDF repo. It means the field name, which should match the "name" field in the json file for plugin UI.

blueandgold commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+  private static final Logger LOG = LoggerFactory.getLogger(AnaplanService.class);
+
+  /**
+   * Sets metadata of Anaplan API root URL and authentication service URL
```
nit: the description here is too tightly coupled to the implementation of the method, and could change in the future.
So, how about a more appropriate description like "Set the metadata details for XYZ task"?

niceyang replied:
Thanks! Updated.

blueandgold commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+  /**
+   * Sets metadata of Anaplan API root URL and authentication service URL
+   *
+   * @param config the config instance carries metadata the task
```
Something is missing here in the wording? "carries metadata the task"

niceyang replied:
Ops, fixed.
"the config instance carries metadata for the connection with Anaplan server"

blueandgold commented on src/main/java/com/anaplan/client/AnaplanService.java:
```diff
@@ -0,0 +1,228 @@
+    }
+  }
+
+  private static void setCredential(String user, String pass) {
```
Can we use a fullname?  like `password`, and `username`?
They map the method name `.getUsername()` and `getPassword()`

niceyang replied:
Sure, thanks!

blueandgold commented on src/main/java/com/google/planningworks/cdap/plugins/action/AnaplanFunctionRunner.java:
```diff
@@ -0,0 +1,63 @@
+   * throws an exception, the pipeline will not be deployed and the user will be shown the error
+   * message.
+   *
+   * @param pipelineConfigure the collection of the pipeline configuration and context
```
This should be called `pipelineConfigurer`

niceyang replied:
Thanks, fixed

blueandgold commented on src/test/java/com/anaplan/client/AnaplanServiceTest.java:
```diff
@@ -0,0 +1,88 @@
+
+  @Test
+  public void validateAuthServiceLocationPass() {
+    AnaplanService.validateAuthServiceLocation("https://google.anaplan.com");
```
Is this a real URL?  If so, can we use a fake one?

blueandgold replied:
Otherwise, drop this test.

blueandgold commented on src/test/java/com/google/planningworks/cdap/plugins/action/AnaplanFunctionRunnerTest.java:
```diff
@@ -0,0 +1,149 @@
+    FieldSetter.setField(
+        config,
+        AnaplanFunctionRunnerConfig.class.getDeclaredField(NAME_WORKSPACE_ID),
+        "8a81b01068d4b6820169da807d121dtt");
```
Are these real values?

blueandgold commented on src/test/java/com/google/planningworks/cdap/plugins/sink/AnaplanSinkTest.java:
```diff
@@ -0,0 +1,124 @@
+    FieldSetter.setField(
+        config,
+        AnaplanSinkConfig.class.getDeclaredField(NAME_AUTH_SERVICE_LOCATION),
+        "https://google.anaplan.com");
```
ditto, drop real value

blueandgold replied:
Please check this everywhere

niceyang replied:
Got, thanks!

blueandgold commented on widgets/AnaplanExport-batchsource.json:
```diff
@@ -0,0 +1,630 @@
+          "name": "serviceLocation",
+          "widget-attributes": {
+            "width": "large",
+            "default": "https://google.anaplan.com"
```
Why is our value being used here?  Is this json for our use?  or for everyone to use?

niceyang replied:
Yeah, for our convenience. But you are right we should not use it here. dropped.

blueandgold commented on widgets/AnaplanExport-batchsource.json:
```diff
@@ -0,0 +1,630 @@
+      }
+    ]
+  }
+}
```
Need newline?

niceyang replied:
Thanks, added!

blueandgold commented on widgets/AnaplanFunctionRunner-action.json:
```diff
@@ -0,0 +1,91 @@
+          "name": "authServiceLocation",
+          "widget-attributes": {
+            "width": "large",
+            "default": "https://google.anaplan.com"
```
ditto

niceyang replied:
Dropped.

blueandgold commented on widgets/AnaplanSink-sparksink.json:
```diff
@@ -0,0 +1,78 @@
+          "name": "authServiceLocation",
+          "widget-attributes": {
+            "width": "large",
+            "default": "https://google.anaplan.com"
```
ditto

niceyang replied:
Dropped.

blueandgold reviewed (changes requested):
Looks great overall!  Major comment is around some actual values in tests and jsons.

niceyang committed:
Addressed the comments from Henry:
1. Improved the Java doc
2. Fixed the parameter names
3. Drop the real values in the unit test
4. Drop the improper default values
5. Other fixes

blueandgold commented on docs/AnaplanExport-batchsource.md:
```diff
@@ -0,0 +1,139 @@
+The data exported from Anaplan model will be firstly materialized in the given GCS bucket, 
+and the inbound data stream to the CDF pipeline will be sourced from the file.
+
+GCP Credentials
```
Should we be more clear about how the service account should work via auto-detect for dataproc?  Or is this a recommendation only?  Or do we recommend people to use the json key file?

niceyang replied:
I got the GCP Credential and GCS & General Properties session directly from the doc for existing GCS source plugin.

As I know the auto-detect is the default setting and the credential will be fetched in the runtime. The credential to be fetched is from the GCP env that runs this CDF instance. Added this detail.

blueandgold commented on docs/AnaplanExport-batchsource.md:
```diff
@@ -0,0 +1,139 @@
+# Anaplan source
+
+Description
```
Just by reading this file, it's not clear what the all the different properties are referring to?  Should we give more context?

niceyang replied:
The context is added. Thanks!

blueandgold commented on docs/AnaplanExport-batchsource.md:
```diff
@@ -0,0 +1,139 @@
+authorization. Can be set to 'auto-detect' when running on a Dataproc cluster.
+When running on other clusters, the file must be present on every node in the cluster.
+
+* **JSON**: Contents of the service account JSON file.
```
Is this for capturing the actual content of the json key file?

niceyang replied:
Yeah, it is.

blueandgold commented on docs/AnaplanSink-sparksink.md:
```diff
@@ -0,0 +1,27 @@
+ **Server File Name:** The name of the predefined placeholder file name in the model for this data import.
+
+ **Service Location:** The root service location of the Anaplan API.
+ Default: https://google.anaplan.com
```
should not be our value here :)

niceyang replied:
Oh forgot here. Fixed.

blueandgold commented on docs/AnaplanSink-sparksink.md:
```diff
@@ -0,0 +1,27 @@
+ Default: https://google.anaplan.com
+ 
+ **Auth Service Location:** The service location for the authentication.
+ Default: https://google.anaplan.com
```
ditto

niceyang replied:
Fixed.

blueandgold commented on pom.xml:
```diff
@@ -0,0 +1,282 @@
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.google.planningworks.cdap.plugins</groupId>
```
ditto

niceyang replied:
That's the group id of this maven project. Something like the metadata in BUILD file.

blueandgold reviewed (changes requested):
Finished another pass. PTAL

niceyang committed:
Addressed the comments from Henry:
1. Drop the real values in the .md doc
2. Improve the .md doc descriptions

blueandgold reviewed (approved):
This is a great contribution to GCP and Anaplan!

albertshau reviewed (commented):
there are over 13k lines here, can you split this up into multiple pull requests? There can at least be one PR per plugin.

I suggest opening one for the simplest plugin first (probably the action?), as there will probably be some style comments that would apply to everything.

niceyang commented:
> there are over 13k lines here, can you split this up into multiple pull requests? There can at least be one PR per plugin.
> 
> I suggest opening one for the simplest plugin first (probably the action?), as there will probably be some style comments that would apply to everything.

Thanks Albert! Yeah let me split the PR. Actually the major of the files are the Anaplan dependency code.

albertshau commented:
closing since this was broken up into a bunch of separate PRs
